### ISSUE TITLE
Project Feedback

### ISSUE THREAD

**Issue by derpachu** [derpachu/MI-449-SS17-740-js-collect-and-display-information-kPt9Fl#1]
@chrisvfritz Can you take a look at this? It's [hosted here](https://derpachu.github.io/MI-449-SS17-740-js-collect-and-display-information-kPt9Fl/) and meets the following criteria:

- [x] Use [this base HTML](https://gist.github.com/chrisvfritz/1b25d9037fb64e3486ea2df8179f9ce8) (feel free to change it) to build a simple app that generates the HTML for a dating profile [in this format](https://gist.github.com/chrisvfritz/4043b4d043610a2bb9cd26444c3e7a45), using information collected from the user
- [x] Show a preview of the dating profile with rendered HTML (e.g. `<em>hella charming</em>` should appear as _hella charming_)
- [x] Also show the raw, generated HTML inside the `<pre><code></code></pre>` elements (e.g. `<em>hella charming</em>` should appear as `<em>hella charming</em>`)
- [x] Both the preview and raw HTML should update live as the user enters their information
- [x] All JavaScript is in a separate `.js` file
- [x] There are no linting errors in your JavaScript (this will also be expected for all future projects)

is there another way to add text formatting under textContent? Also I noticed there was a slight input delay when loading on file, is that just by nature or is there something up with that?

**Comment by chrisvfritz** [comment]
> is there another way to add text formatting under textContent?

If I understand correctly, then yes. 😅  You can use `\n` to create new lines in a string. Let me know if I'm misunderstanding though!

> Also I noticed there was a slight input delay when loading on file, is that just by nature or is there something up with that?

Interesting! I haven't experienced that or heard of it before, but computers do all kinds of weird stuff. 😛  If it's only happening when loading from file, I'd use `live-server` to avoid it.

---

Before diving into the code, I'm seeing some discrepancies between the target HTML and what we're generating. Looking at them side-by-side, your app is producing this:

``` html
<h1>Hi, my name is : FIRST_NAME&nbspLAST_NAME</h1>
<br>
<p>
  DESCRIBE_YOURSELF_INFO
  <br>
  If you're interested in a date, you can email me at:
  <a href="email" target="_blank">EMAIL_ADDRESS</a>
  <br>
  or give me a call at: 
  <a href="phone" target="_blank">PHONE_NUMBER</a>
</p>
```

And we're looking for this:

``` html
<h1>Hi, my name is FIRST_NAME LAST_NAME!</h1>
<p>DESCRIBE_YOURSELF_INFO</p>
<p>
  If you're interested in a date, you can email me at 
  <a href="mailto:EMAIL_ADDRESS" target="_blank">EMAIL_ADDRESS</a>
  or give me a call at
  <a href="tel:PHONE_NUMBER" target="_blank">PHONE_NUMBER</a>.
</p>
```

Let me know if you need some help seeing the differences or if you have any other questions. 🙂

**Comment by derpachu** [comment]
reformatted, let me know if there needs to be some more specific changes to be made

**Comment by chrisvfritz** [comment]
It's a little closer now, but there are still quite a few differences. 

Note that the number of spaces and new lines in the generated HTML is not important. For example, any of these would be acceptable for the first two elements:

``` html
<h1>Hi, my name is FIRST_NAME LAST_NAME!</h1>
<p>DESCRIBE_YOURSELF_INFO</p>
```

``` html
<h1>Hi, my name is FIRST_NAME LAST_NAME!</h1><p>DESCRIBE_YOURSELF_INFO</p>
```

``` html
<h1>Hi, my name is FIRST_NAME LAST_NAME!</h1> <p>DESCRIBE_YOURSELF_INFO</p>
```

But __everything else should be _character-by-character_ identical__, with any data indicated by capitalized words replaced by what is typed into the associated input. 

To verify that what you're producing is correct, you can even type in `FIRST_NAME`, `LAST_NAME`, etc into each `input`, and confirm that what you're producing is same HTML [we provide here](https://gist.github.com/chrisvfritz/4043b4d043610a2bb9cd26444c3e7a45). 

Attention to detail is _vital_ in web development. Please don't hesitate to let me know if you're still not sure how to confirm whether what you're producing is the correct HTML. 🙂

**Comment by derpachu** [comment]
I rewrote it and made the HTML identical except for spacing. Let me know if I missed anything

**Comment by chrisvfritz** [comment]
The HTML is 90% of the way there now, but not yet identical. Take a closer look at the links.

**Comment by derpachu** [comment]
Think I made it identical now

**Comment by chrisvfritz** [comment]
### Attention to detail

Still missing a period (`.`) after the second link, but closer again! 

Remember, attention to detail is _extremely_ important in web development and programming in general. Every character matters! I'd like to see this improve in future projects, as it not only helps us get to the meat of a code review sooner, but most clients and software teams simply wouldn't tolerate four iterations to correctly read less than 2 tweets worth of text.

If you have a visual/cognitive disability that makes this difficult, reach out to me privately on Slack and I can connect you with some resources to make this work a little easier.

### Descriptive variable names

- I see that you're often trying to find ways to shorten variable names (and attribute values), e.g. with `fNameIn` instead of `firstNameInput`. While concise code certainly has its benefits, shorter often conflicts with simpler! Abbreviating variable names isn't a good practice to get into, as it makes your code more difficult to read and understand. The autocompletion in modern text editors can even finish these variables for you, so it doesn't even save time when writing.

- The name `paragraph` isn't very descriptive for the `generated_profile` element. First, it's not the only paragraph on your page and second, the kind of element it is, is actually irrelevant to its purpose in your JavaScript.

These are just two examples. Take a look at _all_ your variable names and make sure they describe the specific content/function and are easy to understand. 🙂 

### Consistent casing

I appreciate that most of the time, you're consistently using `camelCase` in your JavaScript! ⭐️  `updaterawParagraph` is an exception, so let's fix it.

I'm less opinionated about the style you use in `id`/`class` attributes, but let's stay consistent. I see mostly camelCase in your `id`s, but then snake_case for `generated_profile`.

### Paragraph elements

The paragraph elements you're using for your generated HTML and raw HTML are actually a little problematic:

- It's invalid to have block-level elements such as `h1` and `p` inside of another `p` element in HTML. Chrome seems to handle it pretty well, but other browsers may not handle it as gracefully.

- A `p` element inside of a `code` element is also invalid, which can also cause rendering issues in the browser.

Instead, I'd recommend putting your `id`s on the parent `div` and `code` elements, respectively.

### Combining nearly identical functions

Here's a quick way to simplify your code: 

Your `updateParagraph` and `updaterawParagraph` functions are almost exactly the same. You define the same variables, with the same values, and build the same string of HTML. The only difference is you assign that string to `paragraph.innerHTML` in one, but to `rawPara.textContent` in the other.

You should actually be able to combine these functions and cut your code almost in half by assigning the string of HTML to a variable, then reusing it.

---

Please don't hesitate to let me know if you have any questions. If anything isn't clear, I'm happy to provide more detail/resources. 🙂

**Comment by derpachu** [comment]
Aight fixed the issues with the code. As for the making the stuff identical, I think its just the way I read makes it so I only pick up the major points of the sentence. I think moving forward I'll just write a python script to check to make sure all the characters are exactly the same when needed.

**Comment by chrisvfritz** [comment]
> I think its just the way I read makes it so I only pick up the major points of the sentence.

Unfortunately, that just isn't acceptable in a professional environment. If it's a learning disability, I'd like to work with you to find resources that can help. For example, there are browser extensions that can enlargen and/or highlight words to make them easier to read. If it's just a bad habit, you have to force yourself to read more slowly and carefully. Pythons scripts will only be able to help in a small percentage of cases.

The project is looking better, but there are a couple items you missed:

- Many of the names are much improved, 👍  though I do still see less descriptive shortenings, such as the aforementioned `fNameIn`
- We're missing one step from combining the functions: `assigning the string of HTML to a variable, then reusing it`

Let me know if you have any questions about these.

**Comment by derpachu** [comment]
Yeah I'll take more notice moving forward. Thanks for your patience in this matter. Updated all the variable names. Do variables use typing information or is it just going to be general purpose?

**Comment by chrisvfritz** [comment]
Looks good! 🎉 🌟 🎈  :shipit: 

> Do variables use typing information or is it just going to be general purpose?

Great question. 👍  Variables in JavaScript are dynamic, so they can store _any_ value, similar to Python. And just like in Python, they can also be assigned a value of a different type after they're initially declared, like this:

``` js
var myVariable = 'some string'
myVariable = 42
```

Sometimes types can be really useful though, especially in complex library code, so development tools such as [Flow](https://flowtype.org/) have been developed to add _optional_ types to JavaScript code. I wasn't planning on creating a lesson for it in this course, but I could point you to additional resources if you'd be interested in learning more. 🙂

**Comment by derpachu** [comment]
yeah I like to read up on type information. Used to coding with compiled languages instead of interpreted languages

**Comment by chrisvfritz** [comment]
I probably recommend starting with TypeScript then. It's another of the most popular options for adding types to JavaScript, but as I think about it more, I believe it'll provide a gentler learning curve than Flow. All you should need to get started is:

``` sh
npm install --global typescript
```

This will give you a `tsc` (i.e. _TypeScript Compile_) command in the terminal. Before using it though, rename any JavaScript file to end with `.ts` instead of `.js`, like this one:

``` ts
var getFullName = function (firstName, lastName) {
  return firstName + ' ' + lastName
}

getFullName('Tom', 42)
```

Since TypeScript just adds features to JavaScript, any valid JavaScript will also be valid TypeScript. Then you can start adding types, like these:

``` ts
var getFullName = function (firstName: string, lastName: string): string {
  return firstName + ' ' + lastName
}

getFullName('Tom', 42)
```

Now if you run this in the terminal:

``` sh
tsc name-of-file.ts
```

You should see a compile error:

```
Argument of type '42' is not assignable to parameter of type 'string'.
```

If there are _no_ compile errors, It'll create a normal JavaScript file, with the types and any other TypeScript-specific features removed. In editors like Visual Studio Code (what I use as my primary editor), you'll even see type errors out-of-the-box:

![Visual Studio Code compile error](https://i.imgur.com/wwhagZd.png)

There's probably a plugin for Atom that will do the same. Then once you get setup, I'd recommend going through [the official TypeScript tutorial](https://www.typescriptlang.org/docs/tutorial.html). If you get stuck or have questions at any point, don't hesitate to reach out on Slack. 🙂  Happy typing!
